Clarify model-outputs section (#176)
hubverse-org/hubDocs: docs/source/user-guide/model-output.md

## Directory structure
The `model-output` directory in a modeling hub is required to have the following subdirectory and file structure:

* `team1-modelA`
   * `<round-id1>-<model_id>.csv` (or parquet, etc)
   * `<round-id2>-<model_id>.csv` (or parquet, etc)
* `team1-modelB`
   * `<round-id1>-<model_id>.csv` (or parquet, etc)
* `team2-modelA`
   * `<round-id1>-<model_id>.csv` (or parquet, etc)

where `model_id` = `team_abbr-model_abbr`

Note that file names are also allowed to contain the following compression extension prefixes: .snappy, .gzip, .gz, .brotli, .zstd, .lz4, .lzo, .bz2, e.g. `<round-id1>-<model_id>.gz.parquet`.

(model-output-example-table)=
### Example model submission file

Each model submission file will have the same representation for each hub. Here
is an example for a Hub that collects mean and quantile forecasts for
one-week-ahead incidence, but probabilities for the timing of a season peak:  

> :::{table} An example of a model output submission for modelA

Reviewer: Love the example table here (as long as we're sure that round_id doesn't need to be in YYYY-MM-DD format, which I continue to be confused about)

Author: Thank you! This was lifted directly from further down in the document... I just realized that it would have helped me a lot to see what these tables actually looked like before going into the weeds of what the columns could be.

Author: And I just checked in the schema, it _does not_ need to be in ISO-8601 format.

Author: Oh-ho! https://github.com/hubverse-org/hubValidations/discussions/13

Reviewer: > Oh-ho! [hubverse-org/hubValidations#13](https://github.com/hubverse-org/hubValidations/discussions/13)

I still don't understand where we landed on the YYYY-MM-DD thing. If it's still the case that our tooling expects YYYY-MM-DD, I'd propose that the model output file example use a reference date, origin date, or something else that can expressed in that format.

Reviewer 2: I think we do expect any dates to be in YYYY-MM-DD format.

I think round IDs could in theory be something that are not dates. Would be nice to have @annakrystalli confirm.

Reviewer 3: Yup, currently `round_id`s can consist of alphanumerics and underscores or ISO dates

Author: I will resolve this as the statement has not deviated from the original description.  When we make a decision to change it, an issue should be opened and we will address it there.

:::

### File formats

Hubs can take submissions in tabular data formats, namely `csv` and `parquet`. These
submission formats are _not mutuatlly exclusive_; **Hubs may choose between
 `parquet` (arrow) `csv`, or both**. Both formats have advantages and tradeoffs:

* Considerations about `csv`:
   * Advantages
     * Compatibility: Files are human-readable and are widely supported by many tools
   * Disadvantages:
     * Size: Some projects have run into 100 MB file size limits when using `csv` formatted files.
* Considerations about `parquet`:
   * Advantages:
      * Speed
      * Size: In combination, splitting files up and using `parquet` would get around GitHub limits on file sizes
      * Loads only data that are needed
   * Disadvantages:
      * Compatibility: Harder to work with; teams and people who want to work with files need to install additional libraries

(model-output-format)=
## Formats of model output

```{admonition} Reference
Much of the material in this section has been excerpted and/or adapted from the [hubEnsembles manuscript](https://github.com/hubverse-org/hubEnsemblesManuscript).[^Shandross-etal]
```

> [^Shandross-etal]: Shandross, L., Howerton, E., Contamin, L., Hochheiser, H., Krystalli, A., Consortium of Infectious Disease Modeling Hubs, Reich, N. G., Ray, E.L. (2024). [hubEnsembles: Ensembling Methods in R](https://github.com/hubverse-org/hubEnsemblesManuscript). *(under review for publication)*.

Reviewer 2: Could we have two links here:
1. preprint paper: https://www.medrxiv.org/content/10.1101/2024.06.24.24309416v1
2. the existing link to the github repo?

Author: Absolutely!

```suggestion
[^Shandross-etal]: Shandross, L., Howerton, E., Contamin, L., Hochheiser, H., Krystalli, A., Consortium of Infectious Disease Modeling Hubs, Reich, N. G., Ray, E.L. (2024). [hubEnsembles: Ensembling Methods in R](https://www.medrxiv.org/content/10.1101/2024.06.24.24309416v1). *(under review for publication)* (Repo: https://github.com/hubverse-org/hubEnsemblesManuscript).
```

_Model outputs are a specially formatted tabular representation of predictions._
Each row corresponds to a single, unique prediction and each column provides information about what is being predicted, its scope, and its value. 

> Per hubverse convention, **there are three groups of columns**, each group serving a specific purpose: (1) the **"model ID"** is a single column that denotes which model has produced the prediction, (2) the **"task ID"** columns provide details about what is being predicted, and (3) the three **"model output representation"** columns specifies the type of prediction, identifying information about that prediction, and the value of the prediction. 

Author: Oh this is a very good point and it gets at the whole "personas" aspect.

I might put a side comment about the `model ID` aspect since that's only relevant for people who are consuming the data (admins and analysts)

   1. **"Model ID" (1 column)**: Each model should have a unique identifier
      that is stored in the `model-id` column.
   2. **"Task IDs" (multiple columns)**:  The details of the outcome (the model
      task) are provided by the modeler, and can be stored in a series of "task
      ID" columns as described [in this section on task ID
      variables](#task-id-vars). These "task ID" columns may also include
      additional information, such as any conditions or assumptions that were
      used to generate the predictions. Some example variables include
      `target`, `location`, `reference_date`, and `horizon`.    Although there
      are no restrictions on naming task ID variables, when appropriate, we
      suggest that Hubs adopt the standard task ID or column names and
      definitions specified [in the section on usage of task ID
      variables](#task-id-use).  
   3. **"Model output representation" (3 columns)**: consists of a set of three
      columns specifying how the model outputs are represented. All three of

>       these columns will be used by all Hubs:  

Author: Oh good catch! Definitely a subtle but important distinction.

      1. `output_type` specifies the type of representation of the predictive
         distribution, namely `"mean"`, `"median"`, `"quantile"`, `"cdf"`,
         `"cmf"`, `"pmf"`, or `"sample"`.  
      2. `output_type_id` specifies more identifying information specific to
         the output type, which varies depending on the `output_type`
      3. `value` contains the model’s prediction.  

> The following table provides examples that better explain how the "model output representation" columns are used:  

Author: I agree that the original sentence is vague, but I wonder if this might hit the mark better:  

```suggestion
The following table provides more detail on how to configure the three "model output representation" columns based on the type of prediction used. 
```

> The following table provides examples that better explain how the "model output representation" columns are used:  

Reviewer 3: Mainly fine. I think it's good to stick to the standard name of "output type" not "type of prediction" for clarity. Honestly I feel my proposal was clearer but I'm not too fussed.

* Validation based on rules that can easily be encoded in the JSON schema such as
  ranges of expected values and `output_type_id`s. 

* Validation of more involved rules that cannot be encoded in a json schema are
  implemented separately (such as specific relationships between outputs and
  targets). Examples of such rules include:
   1. A predictive quantile for disease incidence may not be larger than the
      population size in that location.
   2. The probabilities assigned to bins must be non-negative and must sum to 1
      within each model task, up to some specified tolerance.

(model-output-schema)=
## The importance of a stable model output file schema

```{admonition} NOTE
The following discussion addresses two different types of schemas:
 - [hubverse schema](https://github.com/hubverse-org/schemas)---the schema for **validating hub configuration files**
 - [arrow schema](https://arrow.apache.org/docs/11.0/r/reference/Schema.html)---the schema for **model output columns in parquet files**.

This section is primarily a concern for parquet files, which encapsulate a schema within the file, but the broader issues have consequences for all output filetypes.
```

> Model output data are stored as separate files, but we use the `hubData` package to open them as a single [arrow dataset](https://arrow.apache.org/docs/r/reference/Dataset.html).

Reviewer: Not sure if it's worth including here, but even people who don't plan to use `hubData` to interact with model output files will have trouble if the parquet files have different schemas.

_I.e._, uniform schemas aren't only necessary to work with hubverse tools, they're necessary for any parquet-based hub if users want to join model-output files and do analysis across submissions.

Author: It probably is worth adding here in some fashion. I'll add it.

However, there are a number of situations where a single consistent data type cannot be guaranteed, e.g.:
- New rounds introducing changes in custom task ID value data types, which are not covered by the hubverse schema. 
- New rounds introducing changes in task IDs covered by the schema but which accept multiple data types (e.g. `scenario_id` where both `integer` and `character` are accepted or `age_group` where no data type is specified in the hubverse schema).
- Adding new output types, which might introduce `output_type_id` values of a new data type.

While validation of config files will alert hub administrations to discrepancies in task ID value data types across modeling tasks and rounds, modifications that will change the overall data type of model output columns _after submissions have been collected_ could cause downstream issues and _should be avoided_.
Some examples of issues caused by a change in the overal data type of model output columns:
 - data type casting being required in downstream analysis code that used to work, 
 - not being able to filter on columns with data type discrepancies between files before collecting

>  - an inability to open hub model output data as an `arrow` dataset

Reviewer: What if we word this last point to make it clear that `arrow` won't be the only tool impacted by mis-matched parquet schemas?

Something like: 
`- errors when opening hub model output data with popular analytics tools like arrow, pandas, and polars`
?

Author: Definitely a good add.

   * Note that if only point forecasts for the composition were required, a categorical output representation could be used.

> ## Validating forecast values

Author: I'm going to make the executive decision to update this to "prediction" since we use that term earlier in the document.

> Validation of forecast submissions is done with the function 

Author: ```suggestion
Before model outputs can be incorporated into a hub, they must be validated. If a hub is centrally stored on GitHub, validations checks will be automatically performed for each submission (via the [`validate_pr()` function](https://hubverse-org.github.io/hubValidations/reference/validate_submission.html) from the `hubValidations` R package).  

Validation of submissions is done with the function 
```

[`validate_submissions()`](https://hubverse-org.github.io/hubValidations/reference/validate_submission.html)
from the `hubValidations` R package, which performs two validation tasks:

* Validation based on rules that can easily be encoded in the JSON schema such as
  ranges of expected values and `output_type_id`s. 

> * Validation of more involved rules that cannot be encoded in a json schema are
>   implemented separately (such as specific relationships between outputs and
>   targets). Examples of such rules include:
>    1. A predictive quantile for disease incidence may not be larger than the
>       population size in that location.
>    2. The probabilities assigned to bins must be non-negative and must sum to 1
>       within each model task, up to some specified tolerance.

Author: ```suggestion
* Validation of more involved rules that cannot be encoded in a json schema are
  implemented separately (such as specific relationships between outputs and
  targets). You can a table with the details of each check in [the `validate_submission()` documentation](https://hubverse-org.github.io/hubValidations/reference/validate_submission.html#details) and [the `validate_pr()` documentation](https://hubverse-org.github.io/hubValidations/reference/validate_pr.html#details).
```

>       within each model task, up to some specified tolerance.

Author: I agree. It seemed weird to bring up just a couple of things that were checked, but not give a complete list. Your solution is much better!

>       within each model task, up to some specified tolerance.

Author: see https://github.com/hubverse-org/hubDocs/pull/176/files#r1771696626

(model-output-schema)=
## The importance of a stable model output file schema

```{admonition} NOTE
The following discussion addresses two different types of schemas:
 - [hubverse schema](https://github.com/hubverse-org/schemas)---the schema for **validating hub configuration files**

>  - [arrow schema](https://arrow.apache.org/docs/11.0/r/reference/Schema.html)---the schema for **model output columns in parquet files**.

Reviewer 3: ```suggestion
 - [arrow schema](https://arrow.apache.org/docs/11.0/r/reference/Schema.html)---the mapping of model output columns to arrow data types.
```

This does not only involve parquet files. It also refers to the schema created when opening a hub data set using arrow so the original description is more accurate.

This section is primarily a concern for parquet files, which encapsulate a schema within the file, but the broader issues have consequences for all output filetypes.
```

Model output data are stored as separate files, but we use the `hubData` package to open them as a single [arrow dataset](https://arrow.apache.org/docs/r/reference/Dataset.html).[^trouble]
**It is necesssary to ensure that all files conform to the same arrow schema** (i.e. share the same column data types) across the lifetime of the hub.

> When we know that all data types conform to the arrow schema, we can be sure that a hub can be [successfully accessed and fully queryable across all columns as an arrow dataset](https://arrow.apache.org/docs/r/articles/dataset.html)

Reviewer 3: ```suggestion
When we know that all data types conform to the arrow schema, we can be sure that a hub can be [successfully accessed and is fully queryable across all columns as an arrow dataset](https://arrow.apache.org/docs/r/articles/dataset.html)
```

This means that **additions of new rounds _should not_ change the overall hub schema at a later date** (i.e. after submissions have already started being collected). 

> [^troube]: Even if you do not use `hubData` to read model outputs, uniform schemas are still important if you want to join model output files and do analyses across submissions.

Reviewer 2: this should be "trouble" currently it is "troube"

Author: So, what you're telling me is that you are... 

.... looking for trouble ^_^

```suggestion
[^trouble]: Even if you do not use `hubData` to read model outputs, uniform schemas are still important if you want to join model output files and do analyses across submissions.
```

(good catch!)

Many common task IDs should have consistent and stable data types because they are validated during hub configuration.

> However, there are a number of situations where a single consistent data type cannot be guaranteed, e.g.:

Reviewer 3: I actually prefer the original here because it mentions why they are validated and stable (i.e. because they are covered by the schema)

Reviewer 3: Also model outputs are not validated during hub configuration but during file submission so this is a bit confusing as is

Author: I think when I was initially reading this, I was confused by the sentence structure (many W are X, are Y and should Z). I've modified the sentence in https://github.com/hubverse-org/hubDocs/pull/176/files#r1771741223

- New rounds introducing changes in custom task ID value data types, which are not covered by the hubverse schema. 
- New rounds introducing changes in task IDs covered by the schema but which accept multiple data types (e.g. `scenario_id` where both `integer` and `character` are accepted or `age_group` where no data type is specified in the hubverse schema).
- Adding new output types, which might introduce `output_type_id` values of a new data type.

While validation of config files will alert hub administrations to discrepancies in task ID value data types across modeling tasks and rounds, modifications that will change the overall data type of model output columns _after submissions have been collected_ could cause downstream issues and _should be avoided_.

> Some examples of issues caused by a change in the overal data type of model output columns:

Reviewer 3: Again I prefer the original because the description as a range confers the change in severity whereas this is somewhat lost by the bullets

Author: I didn't really understand "These issues can range from" as a range of severity, per se. It just seemed like a range of examples. The only thing that was changed was to move the two sentences of examples to a list. I think this can restore the meaning of the change in severity: 

```suggestion
Changing the overall data type of model output columns can cause a range of issues (in order of _increasing severity_):
```

Reviewer 3: All good!
